File: exercise-2022/simulation-2016/11.cpp
#include <algorithm>
#include <iostream>
#include <iterator>
#include <vector>
#include <unordered_map>
#include <utility>
using namespace std;

const int MAXNUM = 11111;

struct Node
{
    int num, house, size;
    vector<Node *> rel;
};
unordered_map<int, Node *> ind;

struct Home
{
    int minnum, totper;
    double avghou, avgsiz;
};
vector<Home> homes;

void merge(Node *x, int ynum)
{
    Node *y = ind[ynum];
    if (! y)
    {
        y = new Node({ynum, 0});
        ind[ynum] = y;
    }

    x->rel.push_back(y);
    y->rel.push_back(x);
}

unordered_map<int, bool> vis;

void dfs(int num, Home *home)
{
    if (vis[num]) return ;
    vis[num] = true;

    Node *node = ind[num];
    home->totper ++;
    home->avgsiz += node->size;
    home->avghou += node->house;
    home->minnum = min(home->minnum, num);

    for (auto i: node->rel)
        dfs(i->num, home);
}

bool cmp(const Home &i, const Home &j)
{
    if (i.avgsiz != j.avgsiz)
        return i.avgsiz > j.avgsiz;
    else return i.minnum < j.minnum;
}

void solve()
{
    int n;
    cin >> n;
    for (int i=0; i<n; i++)
    {
        int num, father, mother, k;
        cin >> num >> father >> mother >> k;
        
        Node *node = ind[num];
        if (! node)
        {
            node = new Node({num});
            ind[num] = node;
        }

        if (father != -1) merge(node, father);
        if (mother != -1) merge(node, mother);
        
        for (int j=0; j<k; j++)
        {
            int kid;
            cin >> kid;
            merge(node, kid);
        }
        int house, size;
        cin >> house >> size;

        node->house = house;
        node->size = size;
    }

    for (auto i: ind)
        if (! vis[i.first])
        {
            Home home = {MAXNUM};
            dfs(i.first, &home);
            home.avghou /= home.totper;
            home.avgsiz /= home.totper;
           
            homes.push_back(home);
        }
    cout << homes.size() << endl;
    sort(homes.begin(), homes.end(), cmp);
    for (auto i: homes)
    {
        printf("%04d %d %.3f %.3f\n",
            i.minnum, i.totper, i.avghou, i.avgsiz);
    }  
}

int main()
{
    solve();
    return 0;
}

File: 2016/simulation/11.py
import functools
MAXNUM = 11111
class Pers:
    def __init__(self, num, hou=0, siz=0):
        self.num = num
        self.hou = hou
        self.siz = siz
        self.rel = []
class Home:
    def __init__(self):
        self.minn = MAXNUM
        self.pnum = 0
        self.avgh = 0
        self.avgs = 0
ind = {}
vis = set()
homes = []
def merge(num, hou, siz, *rel):
    if num not in ind:
        ind[num] = Pers(num, hou, siz)
    else:
        ind[num].hou = hou
        ind[num].siz = siz
    for i in rel:
        if i == -1:
            continue
        if i not in ind:
            ind[i] = Pers(num=i)
        ind[num].rel.append(i)
        ind[i].rel.append(num)
def dfs(num, home):
    if num in vis:
        return
    vis.add(num)
    home.pnum += 1
    home.avgh += ind[num].hou
    home.avgs += ind[num].siz
    home.minn = min(home.minn, num)
    for i in ind[num].rel:
        dfs(i, home)
def mycmp(self, other):
    if self.avgs > other.avgs:
        return -1
    elif self.minn < other.minn:
        return -1
    else:
        return 1
n = int(input())
for _ in range(n):
    l = list(map(int, input().split(' ')))
    merge(l[0], l[-2], l[-1], l[1], l[2], *l[4:-2])
for i in ind.keys():
    if i not in vis:
        home = Home()
        dfs(i, home)
        homes.append(home)
for i in homes:
    i.avgh /= i.pnum
    i.avgs /= i.pnum
homes.sort(key=functools.cmp_to_key(mycmp))
print(len(homes))
for i in homes:
    print("%04d %d %.3f %.3f" %
        (i.minn, i.pnum, i.avgh, i.avgs))